py4j-java/src/main/java/py4j/Protocol.java

```diff
@@ -402,9 +402,16 @@ public static Object getPythonProxy(String commandPart, Gateway gateway) {
 			throw new Py4JException("Invalid Python Proxy.");
 		}
 
+	    ClassLoadingStrategy classLoadingStrategy = gateway.getClassLoadingStrategy();
+	    if (classLoadingStrategy == null) 
+	        classLoadingStrategy = ReflectionUtil.getClassLoadingStrategy();
+	    // Just in case
+	    if (classLoadingStrategy == null)
```

Comment:
Also, looking at the code, `ReflectionUtil.classForName` and `ReflectionUtil.getClassLoader` is used in more places this this one (see `ReflectionEngine.getConstructor/getField/getMethod`, `TypeUtil.forName/getClass/isInstanceOf`, `HelpPageCommand.getHelpClass`). 

There are many strategies (e.g., overload `ReflectionUtil.classForName` with an additional classForName method that takes a Gateway instance as a second parameter), but methods in `ReflectionUtil` and `TypeUtil` will be the most difficult to refactor because they don't currently have access to the gateway.